### PR TITLE
dsumaroka/jv-array-list

### PR DESCRIPTION
**Pull request by dsumaroka**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,122 @@
...
+            T[] increasedElementData = (T[]) new Object[newCapacity];
+            System.arraycopy(this.elementData, 0, increasedElementData, 0, size);
+            this.elementData = increasedElementData;
+        } else {
+            elementData = (T[]) new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+        }
     }
 }
+
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:122]: each file should end with exactly one empty line

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,122 @@
...
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = indexOf(element);
+        if (index != -1) {
+            return remove(index);
+        }
+        throw new NoSuchElementException("Element not found");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:66]: add element to the message

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,122 @@
...
+    }
+
+    private void indexCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Non existed position");
+        }
+    }
+
+    private void checkRange(int index) {
+        if (index < 0 || index >= elementData.length || index > size) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:86]:
```suggestion
        if (index < 0 || index > size) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,122 @@
...
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void indexCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Non existed position");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:81]: include index and size in the message

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,122 @@
...
+        }
+    }
+
+    private void grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            int newCapacity = oldCapacity + (minCapacity - oldCapacity) + (oldCapacity >> 1);
+            T[] increasedElementData = (T[]) new Object[newCapacity];
+            System.arraycopy(this.elementData, 0, increasedElementData, 0, size);
+            this.elementData = increasedElementData;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:115]:
```suggestion
            elementData = increasedElementData;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,122 @@
...
+                    || elementData[i] != null && elementData[i].equals(element)) {
+                index = i;
+                break;
+            }
+        }
+        return index;
+    }
+
+    private void grow() {
+        if (size == this.elementData.length) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:104]:
```suggestion
        if (size == elementData.length) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,122 @@
...
+    private void grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            int newCapacity = oldCapacity + (minCapacity - oldCapacity) + (oldCapacity >> 1);
+            T[] increasedElementData = (T[]) new Object[newCapacity];
+            System.arraycopy(this.elementData, 0, increasedElementData, 0, size);
+            this.elementData = increasedElementData;
+        } else {
+            elementData = (T[]) new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:118]:
```suggestion
        int oldCapacity = elementData.length;
        int newCapacity = oldCapacity + (minCapacity - oldCapacity) + (oldCapacity >> 1);
        T[] increasedElementData = (T[]) new Object[newCapacity];
        System.arraycopy(this.elementData, 0, increasedElementData, 0, size);
        elementData = increasedElementData;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,122 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:7]:
```suggestion
```

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (okuzan) [reply src/main/java/core/basesyntax/ArrayList.java:104]: don't use `this` when it's optional

**Comment by dsumaroka** [comment]
Oh, sorry. I'll do it. Thank you)


Пн, 4 дек. 2023 г. в 14:40, Oleh Kuzan ***@***.***>:

**Pull request closed without merging**